### PR TITLE
Fixed missing run issue on production

### LINKED ISSUE #4064: Dashboard: missing current run
Several users are having a problem where the current course isn't being returned by the dashboard API. 

See https://odl.zendesk.com/agent/tickets/19315 and https://odl.zendesk.com/agent/tickets/18808

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
fixes https://odl.zendesk.com/agent/tickets/19315
fixes #4064

#### What's this PR do?
- There was two runs sorted reverse/decending. 
  - First run is future and some how enrolled and paid and status of second run is [CourseStatus.WILL_ATTEND](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L388)
  - Second run is currently enrolled and paid 
- Because of this current run was skipped.

#### How should this be manually tested?
**Run command:** `./scripts/test/run_snapshot_dashboard_states.sh --match create_paid_enrolled_currently_with_future_run` on my changes and on master

##### Before 
```
                    "runs": [
                        {
                            "id": 20,
                            "course_id": "course:/v1/iure-animi-eveniet",
                            "title": "CourseRun Minus eveniet qui saepe voluptatum.",
                            "status": "will-attend",
                            "has_paid": true,
                            "position": 1,
                            "course_start_date": "2018-08-05T17:16:03.273534Z",
                            "course_end_date": "2018-09-04T17:16:03.273534Z",
                            "fuzzy_start_date": "Starting Repudiandae nostrum quas delectus animi doloremque.",
                            "enrollment_url": "https://smith-brown.info/",
                            "current_grade": null
                        }
                    ]
```

##### After 
```
                    "runs": [
                        {
                            "id": 19,
                            "course_id": "course:/v0/non-facilis-ratione",
                            "title": "CourseRun Fugit recusandae nostrum dolorem minima cupiditate.",
                            "status": "currently-enrolled",
                            "has_paid": true,
                            "position": 2,
                            "course_start_date": "2018-07-12T11:47:22.554646Z",
                            "course_end_date": "2018-08-11T11:47:22.554646Z",
                            "fuzzy_start_date": "Starting Quia neque reprehenderit quae doloremque.",
                            "enrollment_url": "https://johnson-dixon.biz/",
                            "current_grade": null
                        },
                        {
                            "id": 20,
                            "course_id": "course:/v1/iure-animi-eveniet",
                            "title": "CourseRun Minus eveniet qui saepe voluptatum.",
                            "status": "will-attend",
                            "has_paid": true,
                            "position": 1,
                            "course_start_date": "2018-08-06T11:47:22.686236Z",
                            "course_end_date": "2018-09-05T11:47:22.686236Z",
                            "fuzzy_start_date": "Starting Repudiandae nostrum quas delectus animi doloremque.",
                            "enrollment_url": "https://smith-brown.info/",
                            "current_grade": null
                        }
 ]
```
@pdpinch 
#### Screenshots (if appropriate)
dashboard_state_175_create_paid_enrolled_currently_with_future_run

### CODE REVIEW

**Comment by pdpinch** [comment]
@annagav can you review?

I don't understand the screenshot. Is that before or after your change?

**Comment by amir-qayyum-khan** [comment]
@pdpinch image is showing future run, which is also a bug?
because he paid and enrolled future run dashboard is showing that instead of current run

**Comment by amir-qayyum-khan** [comment]
@pdpinch @annagav I made a change in PR, Now dashboard api will prepend current run. To fix an issue were user has paid and enroll in both current run and future/upcoming run

File: dashboard/api.py [hunk 1]
```diff
@@ -246,14 +246,17 @@ def get_info_for_course(course, mmtrack):
...
-                position=len(course_data['runs']) + 1
-            )
+        formatted_run = format_courserun_for_dashboard(
+            run,
+            status,
+            mmtrack=mmtrack_,
+            position=len(course_data['runs']) + 1
         )
+        if run.is_current:
+            # prepend run on the top
```
Comment (pdpinch) [inline dashboard/api.py]: please add a comment about why this is necessary

**Comment by pdpinch** [comment]
When was CURRENTLY_ENROLLED added as a course status? I'm trying to understand why it was missing. 

Also, are there any unit tests that need to be updated for this? I appreciate the additional Selenium test, but those don't really prevent regressions.

**Comment by amir-qayyum-khan** [comment]
@pdpinch 

> When was CURRENTLY_ENROLLED added as a course status? I'm trying to understand why it was missing.

- He enrolled and paid for course course-v1:MITx+14.310x+2T2018 so status is [CURRENTLY_ENROLLED](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L391)
- He enrolled and paid for future run course-v1:MITx+14.310x+3T2018. So status is  [WILL_ATTEND](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L392)
- Runs are sorted on the base or end date. [Descending Order](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L265)
- Because his latest course run has status `WILL_ATTEND`. it is picked in [loop](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L268)
- And the future run process [here](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L304)
- And first run is left out in second [loop](https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L318)


> Also, are there any unit tests that need to be updated for this? I appreciate the additional Selenium test, but those don't really prevent regressions.

Selenium test is creating the scenario. No unit test is failing. But we can create a unit test to assert current run is on the top

**Comment by amir-qayyum-khan** [comment]
2nd option is to add `CourseStatus.WILL_ATTEND` to https://github.com/mitodl/micromasters/blob/master/dashboard/api.py#L269. Which will automatically pick current run. but this can have ripple in case he has passed run and will attend. It will show passed run (older instead of up coming)

```
# pick the first `not enrolled` or the first
    for run_status in run_statuses:
        if run_status.status not in [CourseRunStatus.NOT_ENROLLED, CourseRunStatus.PAID_BUT_NOT_ENROLLED, CourseStatus.WILL_ATTEND]:
            break
```

**Comment by amir-qayyum-khan** [comment]
@annagav any suggestion? Should i just revert my current run first logic or there is something else in your mind?

**Comment by annagav** [comment]
Let's say there is a current and future course run, and the user is enrolled in the future one. Then which run should be at the top? The current run that the user is not enrolled in or the future one?

**Comment by amir-qayyum-khan** [comment]
@annagav i am limiting this to enroll and paid user. updated code

**Comment by annagav** [comment]
Could you also add a test where user is enrolled in future run but not in current?

**Comment by annagav** [comment]
@amir-qayyum-khan, I spoke to @pdpinch and he is going to look at the tickets one more time to decide if we always want to prioritize the currently enrolled run over future enrolled run.

**Comment by pdpinch** [comment]
Thanks, I've taken a deeper look at this and how it would affect the several tickets we've received on this issue. I think this approach will work as long as, when the user has paid for a future run, we still consider the course to have been paid.

**Comment by amir-qayyum-khan** [comment]
@pdpinch we are checking `mmtrack_.is_enrolled_mmtrack(run.edx_course_key)` which check both enroll and paid

**Comment by amir-qayyum-khan** [comment]
some how celery is crashing

File: dashboard/api.py [hunk 2]
```diff
@@ -246,14 +246,19 @@ def get_info_for_course(course, mmtrack):
...
-                mmtrack=mmtrack_,
-                position=len(course_data['runs']) + 1
-            )
+        formatted_run = format_courserun_for_dashboard(
+            run,
+            status,
+            mmtrack=mmtrack_,
+            position=len(course_data['runs']) + 1
         )
+        if run.is_current and formatted_run['has_paid']:
```
Comment (amir-qayyum-khan) [inline dashboard/api.py]: travis ci is crashing on  add`mmtrack_.is_enrolled` check. Any idea @annagav @pdpinch

**Comment by pdpinch** [comment]
What's going on with this PR?

**Pull request merged**